_posts/2023-06-22-a-simple-rest-api-with-play-framework-and-scala.md

```diff
@@ -0,0 +1,420 @@
+
+
+
+  override def save(carDTO: CarDTO): Either[ManagerError, CarDTO] =
+    repository
+      .save(carDTO.toDB())
+      .fold(error => Left(toManagerError(error)), car => Right(car.toDTO()))
+
+
+  override def update(carDTO: CarDTO): Either[ManagerError, Unit] =
+    repository
+      .update(carDTO.toDB())
+      .fold(error => Left(toManagerError(error)), _ => Right(()))
+
+  override def delete(carDTO: CarDTO): Either[ManagerError, Unit] =
+    repository
+      .delete(carDTO.id.getOrElse(0L))
+      .fold(error => Left(toManagerError(error)), _ => Right(()))
+
+  override def getAll(): Either[ManagerError, List[CarDTO]] =
+    repository
+      .findAll()
+      .fold(error => Left(toManagerError(error)), carList => Right(carList.map(_.toDTO())))
+
+  private def toManagerError(repositoryError: RepositoryError) = ManagerError(repositoryError.message)
+
+}
+```
+
+## 3.3 Controllers
+This is the layer where we define ouρ APIs public endpoints.
+
+```scala
+package controllers
+
+import controllers.actions.CarActions
+import play.api.mvc.{BaseController, ControllerComponents, PlayBodyParsers}
+import managers.CarManager
+import models.CarDTO
+import play.api.libs.json.Json
+
+import scala.concurrent.ExecutionContext
+
+class HomeController(val carManager: CarManager, val controllerComponents: ControllerComponents)(implicit val ec: ExecutionContext)
+  extends BaseController
+    with CarActions{
+
+  def get(id: Long) = Action{
+    carManager
+      .get(id)
+      .fold(
+        error => BadRequest(error.message),
+        result => result.map(car => Ok(Json.toJson(car))).getOrElse(NoContent)
+      )
+  }
+
+  def getAll() = Action {
+    carManager
+      .getAll()
+      .fold(
+        error => BadRequest(error.message),
+        result => Ok(Json.toJson(result))
+      )
+  }
+
+  def update() = Action(parse.json[CarDTO]){ implicit request =>
+      carManager
+      .update(request.body)
+      .fold(
+        error => BadRequest(error.message),
+        _ => NoContent
+      )
+  }
+
+  def delete(id: Long) = CarAction(id){ implicit request =>
+    request.carDTO.map(car =>
+    carManager
+      .delete(car)
+      .fold(
+        error => BadRequest(error.message),
+        _ => NoContent
+      )
+    ).getOrElse(BadRequest)
+  }
+
+  def save() = Action(parse.json[CarDTO]){ implicit request =>
+    carManager
+      .save(request.body)
+      .fold(
+        error => BadRequest(error.message),
+        car => Ok(Json.toJson(car))
+      )
+  }
+
+  def index() = Action{
+    Ok("This is a test")
+  }
+
+  override def playBodyParsers: PlayBodyParsers = controllerComponents.parsers
+}
+```
+
+We use the Action Component that is a function of type Request => Result. The component receives an http request and after we pass this through all our layers and transformations our application responts with a valid http response or a respective error code with an informative message.
```

Comment:
Need to explain the structure of a request-response flow first. Then introduce the concept of an action, then explain the controller. Otherwise the code of the controller becomes confusing, and the reader will wtf-throw-table at the sight of an unintroduced concept directly in the code.